src/transformers/models/t5/modeling_t5.py

```diff
@@ -945,23 +952,61 @@ def forward(
             if output_hidden_states:
                 all_hidden_states = all_hidden_states + (hidden_states,)
 
-            layer_outputs = layer_module(
-                hidden_states,
-                attention_mask=extended_attention_mask,
-                position_bias=position_bias,
-                encoder_hidden_states=encoder_hidden_states,
-                encoder_attention_mask=encoder_extended_attention_mask,
-                encoder_decoder_position_bias=encoder_decoder_position_bias,
-                layer_head_mask=layer_head_mask,
-                encoder_layer_head_mask=encoder_layer_head_mask,
-                past_key_value=past_key_value,
-                use_cache=use_cache,
-                output_attentions=output_attentions,
-            )
+            if (
+                getattr(self.config, "gradient_checkpointing", False)
+                and self.training
+                and (not self.config.is_decoder)
```

Comment:
I guess it could/should also work for the decoder no?

Reply:
For the encoder, we only need to deal with the None return value (`present_key_value_state`). Things are a bit more complicated with the decoder. Sometimes some of the returned tensors have `requires_grad=False`. I don't think there is a perfect way to deal with that before PyTorch 1.8. If we enable gradients for those tensors, there will be some unnecessary computation. I also don't know how to do that yet. IMO, we should use another PR to deal with the decoder.

Reply:
I've tracked down the tensor that is causing the problem. It's the position bias returned by `T5Attention` with `has_relative_attention_bias=False`. The gradient checkpointing would work if we force the position bias to require gradients when training ([as in this commit](https://github.com/veritable-tech/transformers/commit/acaeee6b2e675045fb28ce2176444c1d63e908bd)). I'm not sure about the ramification of this change. If it seems okay to you, I can incorporate this commit into this PR.

Reply:
We can also only apply this rule only when gradient checkpointing is on ([as in this commit](https://github.com/veritable-tech/transformers/commit/658bdd0bd1215353a8770f558bda2ea69a0ad0c7)).

Reply:
Thanks for investigating! 

Adding this change:

```python
                if self.training and self.gradient_checkpointing:
                    position_bias.requires_grad = True
```

sounds great to me!

Reply:
Awesome! I've incorporated the changes in this PR.